orc8r/cloud/go/services/bootstrapper/servicers/registration/cloud_registration_test.go

```diff
@@ -0,0 +1,124 @@
+		}
+	)
+
+	factory := test_utils.NewSQLBlobstore(t, bootstrapper.BlobstoreTableName)
+	s := registration.NewBlobstoreStore(factory)
+
+	c, err := registration.NewCloudRegistrationServicer(s, exampleRootCA, exampleTimeout)
+	assert.NoError(t, err)
+
+	ctx := context.Background()
+
+	nonce := registration.GenerateNonce(registration.NonceLength)
+	assert.Equal(t, registration.NonceLength, len(nonce))
+	token := registration.NonceToToken(nonce)
+	assert.Equal(t, nonce, registration.NonceFromToken(token))
+
+	// Try getting device info when token is invalid
+	getGatewayDeviceInfoRes, err := c.GetGatewayDeviceInfo(ctx, &protos.GetGatewayDeviceInfoRequest{Token: token})
+	assert.NoError(t, err)
+	assert.Equal(t, &protos.GetGatewayDeviceInfoResponse_Error{
+		Error: fmt.Sprintf("could not get token info from token %v: %v", token, "Not found"),
+	}, getGatewayDeviceInfoRes.Response)
+
+	// Register some device info
+	getTokenRes, err := c.GetToken(ctx, &protos.GetTokenRequest{
+		GatewayDeviceInfo: gatewayDeviceInfo,
+		Refresh: false,
+	})
+	assert.NoError(t, err)
+	assert.True(t, clock.Now().Before(time.Unix(getTokenRes.Timeout.Seconds, int64(getTokenRes.Timeout.Nanos))))
+	token = getTokenRes.Token
+	timeout := getTokenRes.Timeout
+
+	// Get device info
+	getGatewayDeviceInfoRes, err = c.GetGatewayDeviceInfo(ctx, &protos.GetGatewayDeviceInfoRequest{Token: token})
+	assert.NoError(t, err)
+	assert.Equal(t, &protos.GetGatewayDeviceInfoResponse_GatewayDeviceInfo{GatewayDeviceInfo: gatewayDeviceInfo}, getGatewayDeviceInfoRes.Response)
+
+	// Refresh device info
+	getTokenRes, err = c.GetToken(ctx, &protos.GetTokenRequest{
+		GatewayDeviceInfo: gatewayDeviceInfo,
+		Refresh: true,
+	})
+	assert.NoError(t, err)
+	assert.True(t, time.Unix(timeout.Seconds, int64(timeout.Nanos)).Before(time.Unix(getTokenRes.Timeout.Seconds, int64(getTokenRes.Timeout.Nanos))))
+	assert.NotEqual(t, token, getTokenRes.Token)
+
+	// Get device info with new token
+	getGatewayDeviceInfoRes, err = c.GetGatewayDeviceInfo(ctx, &protos.GetGatewayDeviceInfoRequest{Token: getTokenRes.Token})
+	assert.NoError(t, err)
+	assert.Equal(t, &protos.GetGatewayDeviceInfoResponse_GatewayDeviceInfo{GatewayDeviceInfo: gatewayDeviceInfo}, getGatewayDeviceInfoRes.Response)
+
+	// Old token should still work
+	getGatewayDeviceInfoRes, err = c.GetGatewayDeviceInfo(ctx, &protos.GetGatewayDeviceInfoRequest{Token: token})
+	assert.NoError(t, err)
+	assert.Equal(t, &protos.GetGatewayDeviceInfoResponse_GatewayDeviceInfo{GatewayDeviceInfo: gatewayDeviceInfo}, getGatewayDeviceInfoRes.Response)
+
+	clock.SetAndFreezeClock(t, time.Now().Add(exampleTimeout))
+
+	// Token should be expired
+	getGatewayDeviceInfoRes, err = c.GetGatewayDeviceInfo(ctx, &protos.GetGatewayDeviceInfoRequest{Token: getTokenRes.Token})
+	assert.NoError(t, err)
+	assert.Equal(t, &protos.GetGatewayDeviceInfoResponse_Error{
+		Error: fmt.Sprintf("token %v has expired", getTokenRes.Token),
+	}, getGatewayDeviceInfoRes.Response)}
```

Comment:
Probably other issues -- run `./build.py` / lint
```suggestion
	}, getGatewayDeviceInfoRes.Response)
}
```